Exploit nbdime to show diffs of outputs

The current workflow for nbval is that it a text-based diff of the computed and stored output is presented. Here is an example for failing cells 6 and 17:
<pre>
bin:nbval fangohr$ py.test -v --nbval documentation.ipynb 
============================================ test session starts ============================================
platform darwin -- Python 3.5.2, pytest-2.9.2, py-1.4.31, pluggy-0.3.1 -- /Users/fangohr/anaconda3/bin/python
cachedir: .cache
rootdir: /Users/fangohr/git/nbval, inifile: 
plugins: nbval-0.3.6, cov-2.3.1
collected 10 items 

documentation.ipynb::Cell 6 FAILED
documentation.ipynb::Cell 13 PASSED
documentation.ipynb::Cell 15 PASSED
documentation.ipynb::Cell 17 FAILED
documentation.ipynb::Cell 19 FAILED
documentation.ipynb::Cell 21 FAILED
documentation.ipynb::Cell 24 PASSED
documentation.ipynb::Cell 26 PASSED
documentation.ipynb::Cell 28 FAILED
documentation.ipynb::Cell 29 PASSED

================================================= FAILURES ==================================================
__________________________________________________ cell 6 ___________________________________________________
Notebook cell execution failed
Cell 6: Error with cell

Input:
%%writefile doc_sanitize.cfg
[regex1]
regex: \d{1,2}/\d{1,2}/\d{2,4}
replace: DATE-STAMP

[regex2]
regex: \d{2}:\d{2}:\d{2}
replace: TIME-STAMP

Traceback: mismatch 'text'
<<<<<<<<<<<< Reference output from ipynb file:
Writing doc_sanitize.cfg

============ disagrees with newly computed (test) output:  
Overwriting doc_sanitize.cfg

>>>>>>>>>>>>
__________________________________________________ cell 17 __________________________________________________
Notebook cell execution failed
Cell 17: Error with cell

Input:
print([np.random.rand() for i in range(4)])
print([np.random.rand() for i in range(4)])

Traceback: mismatch 'text'
<<<<<<<<<<<< Reference output from ipynb file:
[0.5484774726012661, 0.6435546033932157, 0.10826743499682889, 0.5748413548528436]
[0.7290940674500538, 0.6663117586823235, 0.7182293584340027, 0.22383996412490337]

============ disagrees with newly computed (test) output:  
[0.4660913775186526, 0.31178898136714595, 0.25151291353922156, 0.5269909473164134]
[0.4953052207494243, 0.04899424112512207, 0.14481404964135336, 0.5252754048311362]

>>>>>>>>>>>>
</pre>

Would it be possible to use [nbdime](https://github.com/jupyter/nbdime) to display those diffs in a nicer way (in the browser)?

If so, then two possible user interfaces come to mind:

1. creation of a diff-document in browser which shows the changes in nbdime style for each failing test. (The changes here are the changes between the saved output cell, and the computed output data. The input data will be the same as it comes from the saved notebook for both outputs.)

2. the option to display the diff output for just a particular cell in the saved document, for example cell 6 in the example above, and to produce the nbdime diff output just for that cell. This might be useful to focus on a particular failing cell. This would require a new flag to py.test to indicated which cell we want to focus on. We would probably still have to execute all notebook cells up to that cell to make sure the commands in cell 6 can reasonably execute.

--- a/nbval/nbdime_reporter.py
+++ b/nbval/nbdime_reporter.py
@@ -23,6 +23,7 @@ from nbdime.webapp.nbdiffweb import run_server, browse
 
 from .plugin import IPyNbCell
 
+nbdime.log.set_nbdime_log_level('ERROR')
 
 _re_nbval_nodeid = re.compile('.*\.ipynb::Cell \d+')
 

--- a/nbval/nbdime_reporter.py
+++ b/nbval/nbdime_reporter.py
@@ -21,7 +21,7 @@ import nbformat
 import nbdime
 from nbdime.webapp.nbdiffweb import run_server, browse
 
-from .plugin import IPyNbCell
+from .plugin import IPyNbCell, bcolors
 
 nbdime.log.set_nbdime_log_level('ERROR')
 
@@ -41,14 +41,17 @@ class NbdimeReporter:
 
         self.stats = {}
 
+    # ---- These functions store captured test items and reports ----
+
     def pytest_runtest_logreport(self, report):
+        """Store all test reports for evaluation on finish"""
         rep = report
         res = self.config.hook.pytest_report_teststatus(report=rep)
         cat, letter, word = res
         self.stats.setdefault(cat, []).append(rep)
 
     def pytest_collectreport(self, report):
-        """Store all collected tests for evalutation on finish
+        """Store all collected nbval tests for evaluation on finish
         """
         items = [x for x in report.result if isinstance(x, IPyNbCell)]
         self.nbval_items.extend(items)
@@ -59,6 +62,8 @@ class NbdimeReporter:
 
     @pytest.hookimpl(hookwrapper=True)
     def pytest_sessionfinish(self, exitstatus):
+        """Called when test session has finished.
+        """
         outcome = yield
         outcome.get_result()
         summary_exit_codes = (
@@ -69,13 +74,18 @@ class NbdimeReporter:
             self.make_report(outcome)
 
     def make_report(self, outcome):
+        """Make report in form of two notebooks.
+
+        Use nbdime diff-web to present the difference between reference
+        cells and test cells.
+        """
         failures = self.getreports('failed')
         if not failures:
             return
         for rep in failures:
             # Check if this is a notebook node
             msg = self._getfailureheadline(rep)
-            self.section(msg)
+            self.section(msg, rep.longrepr.splitlines()[1])
             self._outrep_summary(rep)
         tmpdir = tempfile.mkdtemp()
         try:
@@ -104,9 +114,13 @@ class NbdimeReporter:
                 l.append(x)
         return l
 
-    def section(self, title):
+    def section(self, title, details):
         # Create markdown cell with title
-        header = nbformat.v4.new_markdown_cell("## " + title)
+        source = "## " + title
+        if details:
+            details = details.replace(bcolors.OKBLUE, '')
+            source += "\n\n**" + details + '**'
+        header = nbformat.v4.new_markdown_cell(source)
         # Add markdown in both ref and test
         self.nb_ref.cells.append(header)
         self.nb_test.cells.append(header)
@@ -121,7 +135,8 @@ class NbdimeReporter:
                 ref_cell.outputs = item.sanitize_outputs(ref_cell.outputs)
                 self.nb_ref.cells.append(item.cell)
                 test_cell = copy.copy(item.cell)
-                test_cell.outputs = item.sanitize_outputs(item.test_outputs)
+                if item.test_outputs:
+                    test_cell.outputs = item.sanitize_outputs(item.test_outputs)
                 self.nb_test.cells.append(test_cell)
 
     def _getfailureheadline(self, rep):

--- a/nbval/nbdime_reporter.py
+++ b/nbval/nbdime_reporter.py
@@ -1,0 +1,131 @@
+"""
+pytest ipython plugin modification - Nbdime reporter
+
+Authors: V.T. Fauske
+
+"""
+
+# import the pytest API
+import pytest
+from _pytest.main import EXIT_OK, EXIT_TESTSFAILED, EXIT_INTERRUPTED, \
+    EXIT_USAGEERROR, EXIT_NOTESTSCOLLECTED
+
+import re
+import copy
+import tempfile
+import os
+import shutil
+import io
+
+import nbformat
+import nbdime
+from nbdime.webapp.nbdiffweb import run_server, browse
+
+from .plugin import IPyNbCell
+
+
+_re_nbval_nodeid = re.compile('.*\.ipynb::Cell \d+')
+
+
+class NbdimeReporter:
+    def __init__(self, config, file=None):
+        self.config = config
+        self.verbosity = self.config.option.verbose
+        self._numcollected = 0
+
+        self.nbval_items = []
+
+        self.nb_ref = nbformat.v4.new_notebook()
+        self.nb_test = nbformat.v4.new_notebook()
+
+        self.stats = {}
+
+    def pytest_runtest_logreport(self, report):
+        rep = report
+        res = self.config.hook.pytest_report_teststatus(report=rep)
+        cat, letter, word = res
+        self.stats.setdefault(cat, []).append(rep)
+
+    def pytest_collectreport(self, report):
+        """Store all collected tests for evalutation on finish
+        """
+        items = [x for x in report.result if isinstance(x, IPyNbCell)]
+        self.nbval_items.extend(items)
+        self._numcollected += len(items)
+
+
+    # ---- Code below writes up report ----
+
+    @pytest.hookimpl(hookwrapper=True)
+    def pytest_sessionfinish(self, exitstatus):
+        outcome = yield
+        outcome.get_result()
+        summary_exit_codes = (
+            EXIT_OK, EXIT_TESTSFAILED, EXIT_INTERRUPTED, EXIT_USAGEERROR,
+            EXIT_NOTESTSCOLLECTED)
+        if exitstatus in summary_exit_codes:
+            # We had some failures that might need reporting
+            self.make_report(outcome)
+
+    def make_report(self, outcome):
+        failures = self.getreports('failed')
+        if not failures:
+            return
+        for rep in failures:
+            # Check if this is a notebook node
+            msg = self._getfailureheadline(rep)
+            self.section(msg)
+            self._outrep_summary(rep)
+        tmpdir = tempfile.mkdtemp()
+        try:
+            ref_file = os.path.join(tmpdir, 'reference.ipynb')
+            test_file = os.path.join(tmpdir, 'test_result.ipynb')
+            with io.open(ref_file, "w", encoding="utf8") as f:
+                nbformat.write(self.nb_ref, f)
+            with io.open(test_file, "w", encoding="utf8") as f:
+                nbformat.write(self.nb_test, f)
+            run_server(
+                port=0,     # Run on random port
+                cwd=tmpdir,
+                closable=True,
+                on_port=lambda port: browse(
+                    port, ref_file, test_file, None))
+        finally:
+            shutil.rmtree(tmpdir)
+
+    #
+    # summaries for sessionfinish
+    #
+    def getreports(self, name):
+        l = []
+        for x in self.stats.get(name, []):
+            if not hasattr(x, '_pdbshown'):
+                l.append(x)
+        return l
+
+    def section(self, title):
+        # Create markdown cell with title
+        header = nbformat.v4.new_markdown_cell("## " + title)
+        # Add markdown in both ref and test
+        self.nb_ref.cells.append(header)
+        self.nb_test.cells.append(header)
+
+    def _outrep_summary(self, rep):
+        # Find corresponding item
+        for item in self.nbval_items:
+            if item.nodeid == rep.nodeid:
+                # item found, output
+                # Sanitize reference cell
+                ref_cell = item.cell
+                ref_cell.outputs = item.sanitize_outputs(ref_cell.outputs)
+                self.nb_ref.cells.append(item.cell)
+                test_cell = copy.copy(item.cell)
+                test_cell.outputs = item.sanitize_outputs(item.test_outputs)
+                self.nb_test.cells.append(test_cell)
+
+    def _getfailureheadline(self, rep):
+        if hasattr(rep, 'location'):
+            fspath, lineno, domain = rep.location
+            return fspath + '::' + domain
+        else:
+            return "test session"  # XXX?

--- a/nbval/plugin.py
+++ b/nbval/plugin.py
@@ -78,7 +78,6 @@ def pytest_addoption(parser):
 
 
 def pytest_configure(config):
-    config.option.verbose -= config.option.quiet
     if config.option.nbdime:
         from .nbdime_reporter import NbdimeReporter
         reporter = NbdimeReporter(config, sys.stdout)

--- a/nbval/plugin.py
+++ b/nbval/plugin.py
@@ -564,7 +564,7 @@ class IPyNbCell(pytest.Item):
             # The traceback containing the difference in the outputs is
             # stored in the variable comparison_traceback
             raise NbCellError(self.cell_num,
-                              "Error with cell",
+                              "Cell outputs differ",
                               self.cell.source,
                               # Here we must put the traceback output:
                               '\n'.join(self.comparison_traceback))

--- a/nbval/plugin.py
+++ b/nbval/plugin.py
@@ -203,6 +203,7 @@ class IPyNbCell(pytest.Item):
         self.cell_num = cell_num
         self.cell = cell
         self.docompare = docompare
+        self.test_outputs = None
 
     """ *****************************************************
         *****************  TESTING FUNCTIONS  ***************
@@ -225,7 +226,6 @@ class IPyNbCell(pytest.Item):
         return self.fspath, 0, description
 
     def compare_outputs(self, test, ref, skip_compare=('metadata',
-                                                       'image/png',
                                                        'traceback',
                                                        'text/latex',
                                                        'prompt_number',
@@ -312,10 +312,19 @@ class IPyNbCell(pytest.Item):
         for testing in test:
             for key in testing.keys():
                 if key not in skip_compare:
-                    try:
-                        testing_outs[key] += self.sanitize(testing[key])
-                    except:
-                        testing_outs[key] = self.sanitize(testing[key])
+                    if key == 'data':
+                        for data_key in testing[key].keys():
+                            # Filter the keys in the SUB-dictionary again
+                            if data_key not in skip_compare:
+                                try:
+                                    testing_outs[data_key] += self.sanitize(testing[key][data_key])
+                                except:
+                                    testing_outs[data_key] = self.sanitize(testing[key][data_key])
+                    else:
+                        try:
+                            testing_outs[key] += self.sanitize(testing[key])
+                        except:
+                            testing_outs[key] = self.sanitize(testing[key])
 
 
         # The traceback from the comparison will be stored here.
@@ -480,20 +489,21 @@ class IPyNbCell(pytest.Item):
             # to obtain the 'text' and 'image/png' information
             elif msg_type in ('display_data', 'execute_result'):
                 out['metadata'] = reply['metadata']
+                out['data'] = {}
                 for mime, data in six.iteritems(reply['data']):
-                # This could be useful for reference or backward compatibility
-                #     attr = mime.split('/')[-1].lower()
-                #     attr = attr.replace('+xml', '').replace('plain', 'text')
-                #     setattr(out, attr, data)
+                    # This could be useful for reference or backward compatibility
+                    #     attr = mime.split('/')[-1].lower()
+                    #     attr = attr.replace('+xml', '').replace('plain', 'text')
+                    #     setattr(out, attr, data)
 
-                # Return the relevant entries from data:
-                # plain/text, image/png, execution_count, etc
-                # We coul use a mime types list for this (MAYBE)
-                    setattr(out, mime, data)
+                    # Return the relevant entries from data:
+                    # plain/text, image/png, execution_count, etc
+                    # We could use a mime types list for this (MAYBE)
+                    out.data[mime] = data
                 outs.append(out)
 
-                # if msg_type == 'execute_result':
-                #     out.prompt_number = reply['execution_count']
+                if msg_type == 'execute_result':
+                     out.execution_count = reply['execution_count']
 
 
             # if the message is a stream then we store the output
@@ -516,6 +526,9 @@ class IPyNbCell(pytest.Item):
             else:
                 print("unhandled iopub msg:", msg_type)
 
+        # TODO: Only store if comparing with nbdime, to save on memory usage
+        self.test_outputs = outs
+
         # Compare if the outputs have the same number of lines
         # and throw an error if it fails
         # if len(outs) != len(self.cell.outputs):
@@ -536,6 +549,41 @@ class IPyNbCell(pytest.Item):
                               self.cell.source,
                               # Here we must put the traceback output:
                               '\n'.join(self.comparison_traceback))
+
+    def sanitize_outputs(self, outputs, skip_sanitize=('metadata',
+                                                       'traceback',
+                                                       'latex',
+                                                       'prompt_number',
+                                                       'stdout',
+                                                       'stream',
+                                                       'output_type',
+                                                       'name',
+                                                       'execution_count'
+                                                       )):
+        sanitized_outputs = []
+        for output in outputs:
+            sanitized = {}
+            for key in output.keys():
+                if key in skip_sanitize:
+                    sanitized[key] = output[key]
+                else:
+                    if key == 'data':
+                        sanitized[key] = {}
+                        for data_key in output[key].keys():
+                            # Filter the keys in the SUB-dictionary again
+                            if data_key in skip_sanitize:
+                                sanitized[key][data_key] = output[key][data_key]
+                            else:
+                                sanitized[key][data_key] = self.sanitize(output[key][data_key])
+
+                    # Otherwise, just create a normal dictionary entry from
+                    # one of the keys of the dictionary
+                    else:
+                        # Create the dictionary entries on the fly, from the
+                        # existing ones to be compared
+                        sanitized[key] = self.sanitize(output[key])
+            sanitized_outputs.append(nbformat.from_dict(sanitized))
+        return sanitized_outputs
 
     def sanitize(self, s):
         """sanitize a string for comparison.

--- a/nbval/plugin.py
+++ b/nbval/plugin.py
@@ -27,6 +27,7 @@ from nbformat import NotebookNode
 # Kernel for running notebooks
 from .kernel import RunningKernel, CURRENT_ENV_KERNEL_NAME
 
+
 # define colours for pretty outputs
 class bcolors:
     HEADER = '\033[95m'
@@ -68,6 +69,19 @@ def pytest_addoption(parser):
                     help='Force test execution to use a python kernel in '
                          'the same enviornment that py.test was '
                          'launched from.')
+						 
+    term_group = parser.getgroup("terminal reporting")
+    term_group._addoption(
+        '--nbdime', action='store_true',
+        help="view failed nbval cells with nbdime.")
+
+
+def pytest_configure(config):
+    config.option.verbose -= config.option.quiet
+    if config.option.nbdime:
+        from .nbdime_reporter import NbdimeReporter
+        reporter = NbdimeReporter(config, sys.stdout)
+        config.pluginmanager.register(reporter, 'nbdimereporter')
 
 
 def pytest_collect_file(path, parent):

--- a/nbval/plugin.py
+++ b/nbval/plugin.py
@@ -9,6 +9,7 @@ Authors: D. Cortes, O. Laslett, T. Kluyver, H. Fangohr, V.T. Fauske
 import pytest
 import sys
 import re
+import hashlib
 from collections import OrderedDict
 
 # for python 3 compatibility
@@ -368,11 +369,11 @@ class IPyNbCell(pytest.Item):
                                         + bcolors.FAIL
                                         + "<<<<<<<<<<<< Reference output from ipynb file:"
                                         + bcolors.ENDC)
-                self.comparison_traceback.append(reference_outs[key])
+                self.comparison_traceback.append(_trim_base64(reference_outs[key]))
                 self.comparison_traceback.append(bcolors.FAIL
                                         + '============ disagrees with newly computed (test) output:  '
                                         + bcolors.ENDC)
-                self.comparison_traceback.append(testing_outs[str(key)])
+                self.comparison_traceback.append(_trim_base64(testing_outs[str(key)]))
                 self.comparison_traceback.append(bcolors.FAIL
                                         + '>>>>>>>>>>>>'
                                         + bcolors.ENDC)
@@ -635,3 +636,15 @@ def get_sanitize_patterns(string):
     return re.findall('^regex: (.*)$\n^replace: (.*)$',
                          string,
                          flags=re.MULTILINE)
+
+def hash_string(s):
+    return hashlib.md5(s.encode("utf8")).hexdigest()
+
+_base64 = re.compile(r'^(?:[A-Za-z0-9+/]{4})*(?:[A-Za-z0-9+/]{2}==|[A-Za-z0-9+/]{3}=)?$', re.MULTILINE | re.UNICODE)
+
+def _trim_base64(s):
+    """Trim and hash base64 strings"""
+    if len(s) > 64 and _base64.match(s.replace('\n', '')):
+        h = hash_string(s)
+        s = '%s...<snip base64, md5=%s...>' % (s[:8], h[:16])
+    return s

--- a/nbval/plugin.py
+++ b/nbval/plugin.py
@@ -70,7 +70,7 @@ def pytest_addoption(parser):
                     help='Force test execution to use a python kernel in '
                          'the same enviornment that py.test was '
                          'launched from.')
-						 
+
     term_group = parser.getgroup("terminal reporting")
     term_group._addoption(
         '--nbdime', action='store_true',
@@ -90,10 +90,12 @@ def pytest_collect_file(path, parent):
     Collect IPython notebooks using the specified pytest hook
     """
     if path.fnmatch("*.ipynb"):
+        rich_compare = parent.config.option.nbdime is True
         if parent.config.option.nbval:
-            return IPyNbFile(path, parent)
+            return IPyNbFile(path, parent, rich_compare=rich_compare)
         elif parent.config.option.nbval_lax:
-            return IPyNbFile(path, parent, compare_outputs=False)
+            return IPyNbFile(
+                path, parent, rich_compare=rich_compare, compare_outputs=False)
 
 
 
@@ -123,9 +125,23 @@ class IPyNbFile(pytest.File):
     """
     def __init__(self, *args, **kwargs):
         compare_outputs = kwargs.pop('compare_outputs', True)
+        rich_compare = kwargs.pop('rich_compare', False)
         super(IPyNbFile, self).__init__(*args, **kwargs)
         self.sanitize_patterns = OrderedDict()  # Filled in setup_sanitize_patterns()
         self.compare_outputs = compare_outputs
+        self.skip_compare = (
+            'metadata',
+            'traceback',
+            'text/latex',
+            'prompt_number',
+            'stdout',
+            'stream',
+            'output_type',
+            'name',
+            'execution_count'
+            )
+        if not rich_compare:
+            self.skip_compare = self.skip_compare + ('image/png', 'image/jpeg')
 
     def setup(self):
         """
@@ -240,16 +256,10 @@ class IPyNbCell(pytest.Item):
         description = "cell %d" % self.cell_num
         return self.fspath, 0, description
 
-    def compare_outputs(self, test, ref, skip_compare=('metadata',
-                                                       'traceback',
-                                                       'text/latex',
-                                                       'prompt_number',
-                                                       'stdout',
-                                                       'stream',
-                                                       'output_type',
-                                                       'name',
-                                                       'execution_count'
-                                                       )):
+    def compare_outputs(self, test, ref, skip_compare=None):
+        # Use stored skips unless passed a specific value
+        skip_compare = skip_compare or self.parent.skip_compare
+
         # For every different key, we will store the outputs in a
         # single string, in a dictionary with the same keys
         # At the end, every dictionary entry will be compared

--- a/nbval/plugin.py
+++ b/nbval/plugin.py
@@ -423,6 +423,8 @@ class IPyNbCell(pytest.Item):
 
         # This list stores the output information for the entire cell
         outs = []
+        # TODO: Only store if comparing with nbdime, to save on memory usage
+        self.test_outputs = outs
 
         # Now get the outputs from the iopub channel, need smaller timeout
         timeout = 5
@@ -532,6 +534,11 @@ class IPyNbCell(pytest.Item):
             # cell execution. Therefore raise a cell error and pass the
             # traceback information.
             elif msg_type == 'error':
+                # Store error in output first
+                out['ename'] = reply['ename']
+                out['evalue'] = reply['evalue']
+                out['traceback'] = reply['traceback']
+                outs.append(out)
                 traceback = '\n' + '\n'.join(reply['traceback'])
                 raise NbCellError(self.cell_num, "Cell execution caused an exception",
                                   self.cell.source, traceback)
@@ -540,9 +547,6 @@ class IPyNbCell(pytest.Item):
             # should this raise an error?
             else:
                 print("unhandled iopub msg:", msg_type)
-
-        # TODO: Only store if comparing with nbdime, to save on memory usage
-        self.test_outputs = outs
 
         # Compare if the outputs have the same number of lines
         # and throw an error if it fails

--- a/tests/sanitize_defaults.cfg
+++ b/tests/sanitize_defaults.cfg
@@ -12,5 +12,5 @@ regex: \d\d:\d\d
 replace: TIMESTAMP
 
 [Memory addresses]
-regex: 0x[0-9a-fA-F]+
-replace: MEMORY_ADDRESS
+regex: (<[a-zA-Z_][0-9a-zA-Z_.]* at )(0x[0-9a-fA-F]+)(>)
+replace: \1MEMORY_ADDRESS\3

